module/executiondatasync/optimistic_sync/pipeline/pipeline.go

```diff
@@ -0,0 +1,404 @@
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/executiondatasync/optimistic_sync"
+	"github.com/onflow/flow-go/module/irrecoverable"
+)
+
+var (
+	// ErrInvalidTransition is returned when a state transition is invalid.
+	ErrInvalidTransition = errors.New("invalid state transition")
+)
+
+var _ optimistic_sync.Pipeline = (*Pipeline)(nil)
+
+// worker implements a single worker goroutine that processes tasks submitted to it.
+// It supports submission of context-based tasks that return an error.
+// Each error that occurs during task execution is sent to a dedicated error channel.
+// The primary purpose of the worker is to handle tasks in a non-blocking manner, while still allowing the parent thread
+// to observe and handle errors that occur during task execution.
+type worker struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	pool    *workerpool.WorkerPool
+	errChan chan error
+}
+
+// newWorker creates a single worker.
+func newWorker() *worker {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &worker{
+		ctx:     ctx,
+		cancel:  cancel,
+		pool:    workerpool.New(1),
+		errChan: make(chan error, 1),
+	}
+}
+
+// Submit submits a new task for processing, each error will be propagated in a specific channel.
+// Might block the worker if there is no one reading from the error channel and errors are happening.
+func (w *worker) Submit(task func(ctx context.Context) error) {
+	w.pool.Submit(func() {
+		err := task(w.ctx)
+		if err != nil && !errors.Is(err, context.Canceled) {
+			w.errChan <- err
+		}
+	})
+}
+
+// ErrChan returns the channel where errors are delivered from executed tasks.
+func (w *worker) ErrChan() <-chan error {
+	return w.errChan
+}
+
+// StopWait stops the worker pool and waits for all queued tasks to complete.
+// No additional tasks may be submitted, but all pending tasks are executed by workers before this function returns.
+// This function is blocking and guarantees that any error that occurred during the execution of tasks will be delivered
+// to the caller as a return value of this function.
+// Any error that was delivered during execution will be delivered to the caller.
+func (w *worker) StopWait() error {
+	w.cancel()
+	w.pool.StopWait()
+
+	defer close(w.errChan)
+	select {
+	case err := <-w.errChan:
+		return err
+	default:
+		return nil
+	}
+}
+
+// Pipeline implements the Pipeline interface
+type Pipeline struct {
+	log                  zerolog.Logger
+	stateConsumer        optimistic_sync.PipelineStateConsumer
+	stateChangedNotifier engine.Notifier
+	core                 optimistic_sync.Core
+	worker               *worker
+
+	// The following fields are accessed externally. they are stored using atomics to avoid
+	// blocking the caller.
+	state            *atomic.Int32
+	parentStateCache *atomic.Int32
+	isSealed         *atomic.Bool
+	isAbandoned      *atomic.Bool
+	isIndexed        *atomic.Bool
```

Comment:
why do we have those booleans? I find it counter-intuitive that we have multiple sources. My preference would be to consolidate this with [`State`](https://github.com/onflow/flow-go/blob/7eeb063bd8630b4500e1750bd4bd2e4b30df921d/module/executiondatasync/optimistic_sync/state.go#L4-L17)